internal/wasi/fs.go

```diff
@@ -5,100 +5,137 @@ import (
 	"io"
 	"io/fs"
 	"os"
+	"path"
 	"runtime"
+	"sort"
 	"strings"
-
-	"github.com/tetratelabs/wazero/wasi"
+	"time"
 )
 
+// DirFS is a file system that implements fs.FS and wasi.OpenFileFS.
+// It is similar to os.DirFS, but supports file creation by OpenFile.
+// See os.DirFS.
 type DirFS string
 
-func posixOpenFlags(oFlags uint32, fsRights uint64) (pFlags int) {
-	if fsRights&wasi.R_FD_WRITE != 0 {
-		pFlags |= os.O_RDWR
-	}
-	if oFlags&wasi.O_CREATE != 0 {
-		pFlags |= os.O_CREATE
-	}
-	if oFlags&wasi.O_EXCL != 0 {
-		pFlags |= os.O_EXCL
+// OpenFile implements wasi.OpenFileFS.OpenFile.
+func (dir DirFS) OpenFile(name string, flag int, perm os.FileMode) (fs.File, error) {
+	if !fs.ValidPath(name) || runtime.GOOS == "windows" && strings.ContainsAny(name, `\:`) {
+		return nil, &os.PathError{Op: "open", Path: name, Err: os.ErrInvalid}
 	}
-	if oFlags&wasi.O_TRUNC != 0 {
-		pFlags |= os.O_TRUNC
-	}
-	return
+	return os.OpenFile(string(dir)+"/"+name, flag, perm)
 }
 
-func (dir DirFS) OpenWASI(dirFlags uint32, path string, oFlags uint32, fsRights, fsRightsInheriting uint64, fdFlags uint32) (wasi.File, error) {
-	// I'm not sure how to use all these passed flags and rights yet
-	if !fs.ValidPath(path) || runtime.GOOS == "windows" && strings.ContainsAny(path, `\:`) {
-		return nil, &os.PathError{Op: "open", Path: path, Err: os.ErrInvalid}
-	}
-	mode := fs.FileMode(0644)
-	if oFlags&wasi.O_DIR != 0 {
-		mode |= fs.ModeDir
-	}
-	f, err := os.OpenFile(string(dir)+"/"+path, posixOpenFlags(oFlags, fsRights), mode)
-	if err != nil {
-		return nil, err
-	}
-	return f, nil
+// Open implements fs.FS.Open.
+func (dir DirFS) Open(path string) (fs.File, error) {
+	return dir.OpenFile(path, os.O_RDONLY, 0)
 }
 
-type MemFS struct {
-	Files map[string][]byte
-}
+// MemFS is an in-memory file system that implements fs.FS and wasi.OpenFileFS.
+type MemFS map[string]*memFSEntry
 
-func (m *MemFS) OpenWASI(dirFlags uint32, path string, oFlags uint32, fsRights, fsRightsInheriting uint64, fdFlags uint32) (wasi.File, error) {
-	if !fs.ValidPath(path) {
-		return nil, &os.PathError{Op: "open", Path: path, Err: os.ErrInvalid}
+// OpenFile implements wasi.OpenFileFS.OpenFile.
+func (m MemFS) OpenFile(name string, flag int, perm os.FileMode) (fs.File, error) {
+	if !fs.ValidPath(name) {
+		return nil, &os.PathError{Op: "open", Path: name, Err: fs.ErrInvalid}
 	}
 
-	var buf []byte
-	if oFlags&wasi.O_CREATE == 0 {
-		bts, ok := m.Files[path]
-		if !ok {
-			return nil, &os.PathError{Op: "open", Path: path, Err: os.ErrNotExist}
+	dir := &memFSEntry{Entries: m}
+
+	// Path "." indicates m itself as a directory.
+	if name == "." {
+		return newMemDir(dir, "."), nil
+	}
+
+	// Traverse directories.
+	files := strings.Split(name, "/")
+	for _, file := range files[:len(files)-1] {
+		var ok bool
+		dir, ok = dir.Entries[file]
```

Comment:
do you mean to reassign dir on each iteration of this function? since the `dir` is defined before this loop and also used after this loop, it feels a bit off. if this is intentional, maybe add a comment that we want `dir` to be effected on each path